File: extensions/codex/shared/transcript.ts
import type { MessagePhase } from './protocol/MessagePhase';
import type { TurnError } from './protocol/v2/TurnError';
import type { TurnStatus } from './protocol/v2/TurnStatus';
import type { UserInput } from './protocol/v2/UserInput';

export type CodexTranscriptResourcesReadParams = {
  requests: CodexTranscriptResourceRequest[];
  threadId: string;
};

export type CodexTranscriptResourceRequest =
  | {
      includeTailTurns?: number;
      knownRevision?: string;
      type: 'threadTranscript';
    }
  | {
      knownRevision?: string;
      turnId: string;
      type: 'turn';
    }
  | {
      knownRevision?: string;
      segmentId: string;
      turnId: string;
      type: 'workDetails';
    }
  | {
      itemId: string;
      knownRevision?: string;
      turnId: string;
      type: 'workItem';
    }
  | CodexTranscriptSyncRequest
  | CodexWorkGroupRequest
  | CodexWorkEntryDetailRequest;

export const CODEX_TRANSCRIPT_RENDER_PROTOCOL_VERSION = 2 as const;
export const CODEX_TRANSCRIPT_PROJECTION_VERSION = 'turn-render-v2' as const;
export const DEFAULT_TRANSCRIPT_TAIL_TURNS = 24;
export const DEFAULT_TRANSCRIPT_PREPEND_TURNS = 16;
export const MAX_TRANSCRIPT_WINDOW_TURNS = 40;
export const MAX_TRANSCRIPT_KNOWN_TURNS = 80;
export const DEFAULT_WORK_GROUP_ROWS = 200;
export const MAX_WORK_GROUP_ROWS = 256;

export type CodexTranscriptCapabilities = {
  limits: {
    maxGroupRows: number;
    maxKnownTurns: number;
    maxResponseBytes: number;
    maxWindowTurns: number;
  };
  preferredProtocolVersion: 2;
  projectionVersions: {
    2: typeof CODEX_TRANSCRIPT_PROJECTION_VERSION;
  };
  protocolVersions: Array<1 | 2>;
};

export type CodexTranscriptSyncRequest = {
  knownThreadRevision?: string;
  knownTurns?: Array<{
    renderRevision: string;
    turnId: string;
  }>;
  projectionVersion: typeof CODEX_TRANSCRIPT_PROJECTION_VERSION;
  protocolVersion: 2;
  type: 'transcriptSync';
  window:
    | {
        count?: number;
        kind: 'tail';
      }
    | {
        after: number;
        before: number;
        kind: 'around';
        turnId: string;
      }
    | {
        endTurnId: string;
        kind: 'range';
        startTurnId: string;
      };
};

export type CodexWorkGroupRequest = {
  cursor?: string;
  groupId: string;
  knownRevision?: string;
  limit?: number;
  protocolVersion: 2;
  segmentId: string;
  turnId: string;
  type: 'workGroup';
};

export type CodexWorkEntryDetailRequest = {
  groupId: string;
  knownRevision?: string;
  protocolVersion: 2;
  rowId: string;
  segmentId: string;
  turnId: string;
  type: 'workEntryDetail';
};

export type CodexTranscriptSyncResource = {
  activeTurnId: string | null;
  projectionVersion: typeof CODEX_TRANSCRIPT_PROJECTION_VERSION;
  protocolVersion: 2;
  removedTurnIds: string[];
  sessionId: string | null;
  threadId: string;
  threadRevision: string;
  turnOrder: string[];
  turns: CodexTurnRenderResult[];
  window: {
    endIndexExclusive: number;
    hasEarlier: boolean;
    hasLater: boolean;
    startIndex: number;
    turnIds: string[];
  };
};

export type CodexTurnRenderResult =
  | {
      frame: CodexTurnRenderFrame;
      renderRevision: string;
      status: 'ok';
      turnId: string;
    }
  | {
      renderRevision: string;
      status: 'notModified';
      turnId: string;
    }
  | {
      code: 'frameTooLarge' | 'projectionFailed';
      message: string;
      status: 'error';
      turnId: string;
    };

export type CodexTurnRenderFrame = {
  completedAt: number | null;
  durationMs: number | null;
  error: TurnError | null;
  id: string;
  layoutRevision: string;
  renderRevision: string;
  segments: CodexTurnRenderSegment[];
  startedAt: number | null;
  status: TurnStatus;
};

export type CodexTurnRenderSegment =
  | CodexUserMessageSegment
  | CodexAssistantMessageSegment
  | CodexCompactionSegment
  | CodexWorkRenderSegment;

export type CodexWorkRenderSegment = {
  durationMs: number | null;
  id: string;
  layoutRevision: string;
  revision: string;
  state: 'completed' | 'failed' | 'interrupted' | 'running';
  timeline: CodexWorkTimelineEntry[];
  type: 'work';
};

export type CodexWorkTimelineEntry =
  | (CodexWorkItem & { revision: string })
  | {
      groupType: 'activity' | 'files' | 'text' | 'tools';
      hasMoreRows: boolean;
      id: string;
      revision: string;
      rowCount: number;
      status: 'completed' | 'failed' | 'interrupted' | 'running';
      summary?: CodexWorkGroupSummary;
      title: string;
      type: 'group';
    };

export type CodexWorkGroupSummary = {
  commands: number;
  fileNames: string[];
  files: number;
  reads: number;
  searches: number;
  tools: number;
};

export type CodexWorkGroupResource = {
  groupId: string;
  layoutRevision: string;
  nextCursor: string | null;
  revision: string;
  rows: CodexWorkRowSummary[];
  segmentId: string;
  threadId: string;
  title: string;
  turnId: string;
  type: 'activity' | 'files' | 'text' | 'tools';
};

export type CodexWorkRowSummary =
  | {
      command: string | null;
      durationMs: number | null;
      exitCode: number | null;
      hasDetail: boolean;
      id: string;
      kind: CodexWorkActivity['kind'];
      path: string | null;
      revision: string;
      status: string;
      text: string;
      type: 'activity';
    }
  | {
      additions: number;
      deletions: number;
      hasDetail: boolean;
      id: string;
      kind: CodexFileChange['kind'];
      path: string;
      revision: string;
      status: string;
      type: 'fileChange';
    }
  | {
      category: CodexToolRow['category'];
      detailPreview: string | null;
      hasDetail: boolean;
      id: string;
      label: string;
      mediaCount: number;
      revision: string;
      status: string;
      type: 'tool';
    }
  | {
      hasDetail: false;
      id: string;
      revision: string;
      text: string;
      type: 'text';
    };

export type CodexWorkEntryDetailResource = {
  detail:
    | { detail: string | null; output: string | null; type: 'activity' }
    | { diff: string; type: 'fileChange' }
    | {
        detail: string | null;
        media: CodexMediaPreview[];
        result: string | null;
        type: 'tool';
      };
  groupId: string;
  layoutRevision: string;
  revision: string;
  rowId: string;
  segmentId: string;
  threadId: string;
  truncation: {
    originalBytes: number;
    returnedBytes: number;
    truncated: boolean;
  };
  turnId: string;
};

export type CodexTranscriptResourcesReadResponse = {
  resources: CodexTranscriptResourceResult[];
  threadId: string;
};

export type CodexTranscriptResourceResult = {
  key: string;
  reason?: string;
  requestIndex: number;
  revision?: string;
  status: 'ok' | 'notModified' | 'missing' | 'error';
  value?: unknown;
};

export type CodexThreadTranscriptResource = {
  revision: string;
  sessionId?: string | null;
  threadId: string;
  turnOrder: string[];
  turns?: CodexTranscriptTurn[];
};

export type CodexTurnResource = {
  layoutRevision: string;
  revision: string;
  threadId: string;
  turn: CodexTranscriptTurn;
  turnId: string;
};

export type CodexWorkDetailsResource = {
  details: CodexWorkDetails;
  revision: string;
  segmentId: string;
  threadId: string;
  turnId: string;
};

export type CodexWorkItemResource = {
  item: CodexWorkItem;
  itemId: string;
  revision: string;
  threadId: string;
  turnId: string;
};

export type CodexTranscriptTurn = {
  completedAt: number | null;
  durationMs: number | null;
  error: TurnError | null;
  id: string;
  revision: string;
  segments: CodexTranscriptSegment[];
  startedAt: number | null;
  status: TurnStatus;
};

export type CodexTranscriptSegment =
  | CodexUserMessageSegment
  | CodexWorkSegment
  | CodexAssistantMessageSegment
  | CodexCompactionSegment;

export type CodexUserMessageSegment = {
  content: UserInput[];
  id: string;
  isSteering?: boolean;
  revision: string;
  type: 'userMessage';
};

export type CodexAssistantMessageSegment = {
  id: string;
  phase: MessagePhase | null;
  revision: string;
  text: string;
  type: 'assistantMessage';
};

export type CodexWorkSegment = {
  durationMs: number | null;
  hasDetails: boolean;
  id: string;
  layoutRevision?: string;
  revision: string;
  state: 'running' | 'completed' | 'interrupted' | 'failed';
  timeline?: CodexWorkTimelineEntry[];
  type: 'work';
};

export type CodexCompactionSegment = {
  id: string;
  revision: string;
  status: 'compacting' | 'compacted' | 'cancelled';
  type: 'compaction';
};

export type CodexWorkDetails = {
  entries: CodexWorkEntry[];
  itemIds: string[];
  revision: string;
  segmentId: string;
};

export type CodexWorkEntry =
  | {
      id: string;
      itemId: string;
      type: 'message';
    }
  | {
      id: string;
      itemId: string;
      type: 'userMessage';
    }
  | {
      id: string;
      itemId: string;
      type: 'compaction';
    }
  | {
      group: CodexWorkGroupRef;
      id: string;
      type: 'group';
    };

export type CodexWorkGroupRef = {
  id: string;
  itemIds: string[];
  title: string;
  type: 'activity' | 'files' | 'text' | 'tools';
};

export type CodexWorkItem =
  | {
      id: string;
      phase: MessagePhase | null;
      text: string;
      type: 'message';
    }
  | {
      content: UserInput[];
      id: string;
      isSteering?: boolean;
      type: 'userMessage';
    }
  | {
      id: string;
      status: 'compacting' | 'compacted' | 'cancelled';
      type: 'compaction';
    }
  | {
      activity: CodexWorkActivity;
      id: string;
      type: 'activity';
    }
  | {
      files: CodexFileChange[];
      id: string;
      type: 'fileChanges';
    }
  | {
      id: string;
      row: CodexToolRow;
      type: 'tool';
    };

export type CodexWorkGroup =
  | CodexFileWorkGroup
  | CodexActivityWorkGroup
  | CodexTextWorkGroup
  | CodexToolWorkGroup;

export type CodexFileWorkGroup = {
  files: CodexFileChange[];
  id: string;
  title: string;
  type: 'files';
};

export type CodexActivityWorkGroup = {
  activities: CodexWorkActivity[];
  id: string;
  title: string;
  type: 'activity';
};

export type CodexTextWorkGroup = {
  id: string;
  lines: string[];
  title: string;
  type: 'text';
};

export type CodexToolWorkGroup = {
  id: string;
  rows: CodexToolRow[];
  title: string;
  type: 'tools';
};

export type CodexFileChange = {
  additions: number;
  deletions: number;
  diff: string;
  id: string;
  kind: 'added' | 'deleted' | 'edited' | 'moved';
  path: string;
  status: string;
};

export type CodexWorkActivity = {
  command: string | null;
  detail: string | null;
  durationMs: number | null;
  exitCode: number | null;
  id: string;
  kind: 'read' | 'list' | 'search' | 'webSearch' | 'command' | 'approval';
  output: string | null;
  path: string | null;
  status: string;
  text: string;
};

export type CodexToolRow = {
  category: 'browser' | 'generic' | 'image' | 'nodeRepl';
  detail: string | null;
  id: string;
  label: string;
  media: CodexMediaPreview[];
  result: string | null;
  status: string;
};

export type CodexMediaPreview = {
  id: string;
  label: string | null;
  source:
    | {
        type: 'localPath';
        path: string;
      }
    | {
        type: 'uri';
        uri: string;
      };
};

File: extensions/codex/server/src/transcript.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};
pub(crate) const MAX_TAIL_TURNS: usize = 200;
pub(crate) const TRANSCRIPT_RENDER_PROTOCOL_VERSION: u8 = 2;
pub(crate) const TRANSCRIPT_PROJECTION_VERSION: &str = "turn-render-v2";
pub(crate) const DEFAULT_TRANSCRIPT_TAIL_TURNS: usize = 24;
pub(crate) const MAX_TRANSCRIPT_WINDOW_TURNS: usize = 40;
pub(crate) const MAX_TRANSCRIPT_KNOWN_TURNS: usize = 80;
pub(crate) const DEFAULT_WORK_GROUP_ROWS: usize = 200;
pub(crate) const MAX_WORK_GROUP_ROWS: usize = 256;

#[derive(Debug, Clone)]
pub(crate) struct SessionIndex {
    pub(crate) rollback_hidden_turn_ids: Vec<String>,
    pub(crate) session_id: Option<String>,
    pub(crate) visible_turn_ids: Vec<String>,
    pub(crate) turns: HashMap<String, TurnRange>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct TurnRange {
    pub(crate) turn_id: String,
    pub(crate) start_offset: u64,
    pub(crate) end_offset: u64,
    pub(crate) content_hash: u64,
    pub(crate) started_at: Option<i64>,
    pub(crate) completed_at: Option<i64>,
    pub(crate) duration_ms: Option<i64>,
    pub(crate) status: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct ResourcesReadParams {
    pub(crate) thread_id: String,
    pub(crate) requests: Vec<ResourceRequest>,
}

#[derive(Debug, Deserialize)]
#[serde(tag = "type")]
pub(crate) enum ResourceRequest {
    #[serde(rename = "threadTranscript", rename_all = "camelCase")]
    ThreadTranscript {
        known_revision: Option<String>,
        include_tail_turns: Option<usize>,
    },
    #[serde(rename = "turn", rename_all = "camelCase")]
    Turn {
        turn_id: String,
        known_revision: Option<String>,
    },
    #[serde(rename = "workDetails", rename_all = "camelCase")]
    WorkDetails {
        turn_id: String,
        segment_id: String,
        known_revision: Option<String>,
    },
    #[serde(rename = "workItem", rename_all = "camelCase")]
    WorkItem {
        turn_id: String,
        item_id: String,
        known_revision: Option<String>,
    },
    #[serde(rename = "transcriptSync", rename_all = "camelCase")]
    TranscriptSync {
        protocol_version: u8,
        projection_version: String,
        window: TranscriptWindowRequest,
        known_thread_revision: Option<String>,
        #[serde(default)]
        known_turns: Vec<KnownTurnRevision>,
    },
    #[serde(rename = "workGroup", rename_all = "camelCase")]
    WorkGroup {
        protocol_version: u8,
        turn_id: String,
        segment_id: String,
        group_id: String,
        cursor: Option<String>,
        limit: Option<usize>,
        known_revision: Option<String>,
    },
    #[serde(rename = "workEntryDetail", rename_all = "camelCase")]
    WorkEntryDetail {
        protocol_version: u8,
        turn_id: String,
        segment_id: String,
        group_id: String,
        row_id: String,
        known_revision: Option<String>,
    },
}

#[derive(Debug, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub(crate) enum TranscriptWindowRequest {
    Tail {
        count: Option<usize>,
    },
    Around {
        #[serde(rename = "turnId")]
        turn_id: String,
        before: usize,
        after: usize,
    },
    Range {
        #[serde(rename = "startTurnId")]
        start_turn_id: String,
        #[serde(rename = "endTurnId")]
        end_turn_id: String,
    },
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct KnownTurnRevision {
    pub(crate) turn_id: String,
    pub(crate) render_revision: String,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ValidationReport {
    pub(crate) codex_home: String,
    pub(crate) scanned_files: usize,
    pub(crate) threads_with_turns: usize,
    pub(crate) turns_checked: usize,
    pub(crate) work_details_checked: usize,
    pub(crate) duplicate_segment_failures: usize,
    pub(crate) duplicate_message_warnings: usize,
    pub(crate) invalid_user_input_failures: usize,
    pub(crate) missing_work_details_failures: usize,
    pub(crate) rollback_hidden_turn_failures: usize,
    pub(crate) errors: Vec<String>,
}

pub struct ValidationOptions {
    pub limit: usize,
}
